marklogic-data-hub/src/main/java/com/marklogic/hub/impl/DataHubImpl.java

```diff
@@ -1040,10 +1040,41 @@ protected void prepareProjectBeforeUpgrading(HubProject hubProject, String curre
     // only used in test
     public void setHubConfig(HubConfigImpl hubConfig) {
         this.hubConfig = hubConfig;
+        if (this.loadUserModulesCommand != null) {
+            this.loadUserModulesCommand.setHubConfig(hubConfig);
+        }
+        if (this.loadHubModulesCommand != null) {
+            this.loadHubModulesCommand.setHubConfig(hubConfig);
+        }
+        if (this.loadHubArtifactsCommand != null) {
+            this.loadHubArtifactsCommand.setHubConfig(hubConfig);
+        }
+        if (this.loadUserArtifactsCommand != null) {
+            this.loadUserArtifactsCommand.setHubConfig(hubConfig);
+        }
+        if (this.generateFunctionMetadataCommand != null) {
+            this.generateFunctionMetadataCommand.setHubConfig(hubConfig);
+        }
+    }
+
+    // only used in test
+    public HubConfigImpl getHubConfig() {
+        return this.hubConfig;
     }
 
     // only used in test
     public void setVersions(Versions versions) {
         this.versions = versions;
     }
+
+    public Object clone() throws CloneNotSupportedException {
```

Comment:
Is there any benefit to implementing this vs one-ui just constructing a DataHubImpl itself and then populating the fields necessary? 

The first 3 commands in the list only require a HubConfig. I am wondering if a DataHubImpl(HubConfig) constructor would be better here. That constructor would instantiate the first 3 commands. LoadUserModulesCommand has 2 other dependencies, so the caller of that constructor would need to provide an instance of LoadUserModulesCommand itself. 

I think the fact that our Java classes have private fields that depend on Autowired and don't expose dependencies via constructors makes life really hard for anyone that wants to instantiate our classes without Spring. That's the problem you're running into here. I think the way we solve that is by creating those constructors so that it's much easier for a project like one-ui to instantiate these classes itself, because the constructors will make the dependencies plainly visible.

Reply:
I just noticed DataHubImpl has a couple dependencies that aren't used - I'm going to submit a PR to clean that up.